pyomo/solvers/plugins/solvers/mosek_direct.py

```diff
@@ -122,331 +110,293 @@ def license_is_valid():
-         for i in repn.linear_vars]
-
-        new_expr = [list(repn.linear_coefs), indexes, repn.constant]
-
-        qsubi = []
-        qsubj = []
-        qval = []
-        for i, v in enumerate(repn.quadratic_vars):
-            x, y = v
-            qsubj.append(self._pyomo_var_to_solver_var_map[x])
-            qsubi.append(self._pyomo_var_to_solver_var_map[y])
-            qval.append(repn.quadratic_coefs[i]*((qsubi==qsubj)+1))
-            referenced_vars.add(x)
-            referenced_vars.add(y)
-        new_expr.extend([qval, qsubi, qsubj])
-
-        return new_expr, referenced_vars
-
-    def _get_expr_from_pyomo_expr(self, expr, max_degree=2):
-        if max_degree == 2:
-            repn = generate_standard_repn(expr, quadratic=True)
+        'MOSEK does not support expressions of degree {}.'.format(degree))
+        
+        referenced_vars = ComponentSet(repn.linear_vars)
+        indices = [self._pyomo_var_to_solver_var_map[i] 
+                   for i in repn.linear_vars]
+        mosek_arow = [indices,list(repn.linear_coefs),repn.constant]
+
+        if len(repn.quadratic_vars)==0:
+            mosek_qexp = [[],[],[]]
+            return mosek_arow, mosek_qexp, referenced_vars    
         else:
-            repn = generate_standard_repn(expr, quadratic=False)
-
+            q_vars = itertools.chain.from_iterable(repn.quadratic_vars)
+            [referenced_vars.add(q) for q in q_vars]
+            qsubi, qsubj = zip(*[
+                [self._pyomo_var_to_solver_var_map[xi],
+                self._pyomo_var_to_solver_var_map[xj]] 
+                for xi,xj in repn.quadratic_vars])
+            qval = list(map(
+                lambda v,i,j: v*((i==j)+1),repn.quadratic_coefs,qsubi,qsubj))
+            mosek_qexp = [qsubi,qsubj,qval]
+        return mosek_arow, mosek_qexp, referenced_vars
+    
+    def _get_expr_from_pyomo_expr(self,expr,max_degree=2):
+        repn = generate_standard_repn(expr,quadratic=(max_degree==2))
         try:
-            mosek_expr, referenced_vars = self._get_expr_from_pyomo_repn(
-                repn, max_degree)
+            mosek_arow, mosek_qexp, referenced_vars = self._get_expr_from_pyomo_repn(repn,max_degree)
         except DegreeError as e:
             msg = e.args[0]
-            msg += '\nexpr: {0}'.format(expr)
+            msg += '\nexpr: {}'.format(expr)
             raise DegreeError(msg)
-
-        return mosek_expr, referenced_vars
-
+        return mosek_arow, mosek_qexp, referenced_vars
+    
+    def _mosek_vartype_from_var(self, var):
+        if var.is_integer():
+            return self._mosek.variabletype.type_int
+        return self._mosek.variabletype.type_cont
+    
+    def _mosek_bounds(self, lb, ub, constant = 0.0):
+        if lb is None:
+            lb = -float('inf')
+        if ub is None:
+            ub = float('inf')
+        lb -= constant
+        ub -= constant
+        bound_key = (lb!=-float('inf')) + 2*(ub!=float('inf')) + (lb==ub)
```

Comment:
The last case that you mention, i.e. 1< var <2 seems correct to me. For the other two cases, MOSEK will raise an error due to invalid input: putvarbound method will not accept `float('inf')` as a valid input. Furthermore, if a user inputs an upper bound that is smaller than a lower bound (both finite), then it will make the problem Primal infeasible, which will be evident in the log output. So there should be no cases where the problem would give wrong results due to this.

However, as you mentioned in one of your other comments, if this makes it harder to implement pyomo in other appilcations, then I will change this to the method that was used before.

Reply:
I am not sure the previous implementation is better.  My only intent with the comment was there was an weird edge case and it took me longer to figure out what the 1-liner `bound_key = (...` was doing than I would like to admit.

Reply:
I agree that it is not the most transparent implementation. I have put in place of this a cleaner method. Thanks!